tools/python/util/reduced_build_config_parser.py

```diff
@@ -0,0 +1,123 @@
+    The type information is in a map, with 'inputs' and 'outputs' keys. The value for 'inputs' or 'outputs' is a map
+    between the index number of the input/output and the required list of types.
+
+    For example, both the input and output types are relevant to ai.onnx:Cast.
+    Type information for input 0 and output 0 could look like this:
+        `{"inputs": {"0": ["float", "int32_t"]}, "outputs": {"0": ["float", "int64_t"]}}`
+
+    which is added directly after the operator name in the configuration file.
+    e.g.
+        `ai.onnx;12;Add,Cast{"inputs": {"0": ["float", "int32_t"]}, "outputs": {"0": ["float", "int64_t"]}},Concat`
+
+    If for example the types of inputs 0 and 1 were important, the entry may look like this (e.g. ai.onnx:Gather):
+        `{"inputs": {"0": ["float", "int32_t"], "1": ["int32_t"]}}`
+
+    Finally some operators do non-standard things and store their type information under a 'custom' key.
+    ai.onnx.OneHot is an example of this, where 3 type names from the inputs are combined into a string.
+        `{"custom": ["float_int64_t_int64_t", "int64_t_string_int64_t"]}`
+
+    :param config_file: Configuration file to parse
+    :param enable_type_reduction: Set to True to use the type information in the config.
+                                  If False the type information will be ignored.
+                                  If the flatbuffers module is unavailable any type information will be ignored
+                                  as the usage of the type information is via OperatorTypeUsageManager which has a
+                                  dependency on the ORT flatbuffers python schema.
+    :return: required_ops, op_type_usage_manager:
+             Dictionary of domain:opset:[ops] for required operators
+             OperatorTypeUsageManager manager with operator specific type usage information if available. None if
+             type reduction was disabled, or the flatbuffers module is not available.
+    '''
+
+    if not os.path.isfile(config_file):
+        raise ValueError('Configuration file {} does not exist'.format(config_file))
+
+    required_ops = {}
+    op_type_usage_manager = OperatorTypeUsageManager() if enable_type_reduction and have_flatbuffers else None
+
+    with open(config_file, 'r') as config:
+        for line in [orig_line.strip() for orig_line in config.readlines()]:
+            if not line or line.startswith("#"):  # skip empty lines and comments
+                continue
+
+            domain, opset_str, operators_str = [segment.strip() for segment in line.split(';')]
+            opset = int(opset_str)
+
+            # any type reduction information is serialized json that starts/ends with { and }.
+            # type info is optional for each operator.
+            if '{' in operators_str:
+                # parse the entries in the json dictionary with type info
```

Comment:
would it be simpler to use [json.loads](https://docs.python.org/3/library/json.html#json.loads)?

Reply:
There can be multiple ops per line, so you need to find the start and end '{' "}' for an individual op. json.loads will parse the whole line and accumulate entries from multiple ops into a single dictionary.

I use json.loads in TypeUsageProcessor.from_config_entry and would rather leave that there so it's paired with the json.dumps call in to_config_entry.